worlds/_bizhawk/client.py

```diff
@@ -0,0 +1,84 @@
+"""
+A module containing the BizHawkClient base class and metaclass
+"""
+
+
+from __future__ import annotations
+
+import abc
+from typing import TYPE_CHECKING, Any, Dict, Optional, Tuple, ClassVar
+
+from worlds.LauncherComponents import Component, SuffixIdentifier, Type, components, launch_subprocess
+
+if TYPE_CHECKING:
+    from .context import BizHawkClientContext
+else:
+    BizHawkClientContext = object
+
+
+class AutoBizHawkClientRegister(abc.ABCMeta):
+    game_handlers: ClassVar[Dict[str, Dict[str, BizHawkClient]]] = {}
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], namespace: Dict[str, Any]) -> AutoBizHawkClientRegister:
+        new_class = super().__new__(cls, name, bases, namespace)
+        if "system" in namespace:
+            if namespace["system"] not in AutoBizHawkClientRegister.game_handlers:
+                AutoBizHawkClientRegister.game_handlers[namespace["system"]] = {}
+
+            if "game" in namespace:
+                AutoBizHawkClientRegister.game_handlers[namespace["system"]][namespace["game"]] = new_class()
+
+        return new_class
+
+    @staticmethod
+    async def get_handler(ctx: BizHawkClientContext, system: str) -> Optional[BizHawkClient]:
+        if system in AutoBizHawkClientRegister.game_handlers:
+            for handler in AutoBizHawkClientRegister.game_handlers[system].values():
+                if await handler.validate_rom(ctx):
+                    return handler
+
+        return None
+
+
+class BizHawkClient(abc.ABC, metaclass=AutoBizHawkClientRegister):
+    system: ClassVar[str]
+    """The system that the game this client is for runs on"""
```

Comment:
Pokémon Red and Blue can be run on two different systems: GB and SGB. I need to be able to register my game with either of those two systems being reported

Reply:
Hmm. Edge cases all the way down. I can make it a `Union[str, List[str]]` and just add duplicate references I suppose. And you can detect the system through the async BizHawk functions if you need to. I'll take a closer look tomorrow.

Reply:
This is something that will affect all Gameboy games, and is likely a situation unique to Gameboy games, so perhaps just looking for GB system client handlers when SGB is reported would be a good solution?

Reply:
I'm thinking about, like, GBA in NDS, or someone going crazy and implementing a client that works for one game ported to two systems like Twilight Princess (but something on BizHawk supported systems).

Changed in d706c3f. You can now specify either a `str` or a `tuple[str, ...]`. So in this case `("GB", "SGB")`. I'd rather have it explicit for all multi-system games instead of implicitly allowing certain systems to mix just in case there's ever a difference.